### ISSUE TITLE
logarithmic scale training progress

### ISSUE THREAD

**Issue by keorn** [NVIDIA/DIGITS#113]
It would be great to have an option of logarithmic scale on the training graph. Now, something can be seen at the beginning and then it's just a flat line.

**Comment by lukeyeager** [comment]
Yeah, that's a good idea. Someone else has requested that before. Would you like to see it only for the loss or for the accuracy as well? If I made the accuracy axis logarithmic, I might need to display it as "error rate" (approaching 0), instead of "accuracy" (approaching 100). That might throw people off, even though it's really just displaying the same information.

**Comment by keorn** [comment]
I am not sure if there is a need to display "error rate" or "accuracy" during training. The graph is used to evaluate training progress and for that loss(val) along with loss(train) is sufficient. The accuracy along with confusion matrix is only necessary for the test set.
There could be an option to track test accuracy during training, or just save the best one.

**Comment by lukeyeager** [comment]
Well the test set isn't actually used at the moment, but that's another issue (#17).

I don't agree about not showing the accuracy on the validation set. I think it's helpful to look at. If the user wants to remove the accuracy layer from the "TEST" phase (which is just [caffe-speak](https://github.com/BVLC/caffe/issues/1245#issuecomment-74999077) for the validation phase), that's fine. But I'm going to display whatever the user asks for. I could however see an argument for putting the accuracy outputs on the validation set into a different graph to keep the loss outputs separate.

**Comment by keorn** [comment]
Makes sense, didnt't think about the `TRAIN` - `TEST` distinction that is necessary for some nets.
I think a separate graph for results of inference on the validation set is good, it also makes the distinction from training loss clearer. It can stay linear, so values approaching 1 can be used. It could later have an option for displaying components of confusion matrix on validation as well.
I think both graphs should be linear by default, with a switch to logarithmic as a lot of people are not used to logarithmic scales.

**Comment by lukeyeager** [comment]
> I think both graphs should be linear by default, with a switch to logarithmic as a lot of people are not used to logarithmic scales.

Great idea. Do you want to figure out the Chart.js code necessary to do this and open a pull request? I'll get to it eventually if you don't.

**Comment by lukeyeager** [comment]
@gheinrich addressed this in https://github.com/NVIDIA/DIGITS/commit/245285f96e7cac0d805289c20732319df7a17a57 by adding a subchart. Now you can zoom in on a particular part of the graph:

![subchart-example](https://cloud.githubusercontent.com/assets/687269/8882179/dba079fc-31f9-11e5-8765-135a0aae0456.jpg)

Apparently, you can't do a logarithmic scale with C3.js, but this should still be pretty helpful.

**Comment by PoloShock** [comment]
Even though this thread is closed now, I did some work on logarithmic scale a few days ago, so I thought I would post my results here anyway.
Below are my real-world data plotted in a. current solution; and b. proposed solution using logarithmic scale on y axis.
a. http://jsfiddle.net/poloshock/b33djk91/1/
b. http://jsfiddle.net/poloshock/k6dzvasf/2/

Switching between linear and logarithmic scale could be possible. @lukeyeager do you want me to try to implement it in digits or is logarithmic scale not needed/wanted any more? I think it could still have some potential especially in situation when loss functions tend to be more "spiky".

**Comment by gheinrich** [comment]
Hi Poloshock, thanks for the feedback. We did experiment with logarithmic scales in DIGITS as we also believe there is merit in supporting this type of visualization. However we came across some complexities: C3 does not natively support logarithmic scales so you need to pre-process your data to achieve the desired scale however the tooltip will show the processed data, not the original data. This can be somewhat confusing.
There is also the question of how to handle limits, for example if the loss function falls to exactly zero. Plus we don't know a priori the valid range of the loss function (some networks may have negative loss) so you need to first inspect your data to work out the boundaries of the loss function and then you need to e.g. introduce an arbitrary bias to avoid undefined values. This may not generalise well across the full range of models we want to support.
We thought the subgraph provided a decent alternative to logarithmic scales in that it allows the user to zoom in on any area of the graph and since the axes are automatically (linearly) rescaled, this makes it possible to inspect small variations of the data.
Please let us know what you think.

**Comment by PoloShock** [comment]
Hi @gheinrich , as you pointed out, I indeed had to pre-process the data in order to achieve the desired scale however the tooltip can actually show the original data by referencing the original values which are also present in the "data" structure (there was a typo in previous jsfiddle).
For the limits handling, I biased the data by Number.MIN_VALUE in order to support zero values. However, I didn't expect that loss function can fall to negatives. In that case, data would have to be scanned for the minimum value which would have to be subtracted from the Number.MIN_VALUE bias I used.
JSFiddle with negative loss function values: http://jsfiddle.net/poloshock/t6n53m43/1/

With the complexities growing, I think the benefits of logarithmic scale might not be worth it compared to the subgraph option.
